src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -236,103 +246,75 @@ void executeQueryWithParallelReplicas(
-            visitor.visit(query_ast_for_shard);
-        }
-        else
-            query_ast_for_shard = query_ast;
-
-        stream_factory.createForShardWithParallelReplicas(
-            shard_info, query_ast_for_shard, main_table, context, static_cast<UInt32>(shards), plans, remote_shards);
-    }
-
-    Scalars scalars = context->hasQueryContext() ? context->getQueryContext()->getScalars() : Scalars{};
-    scalars.emplace(
-        "_shard_count", Block{{DataTypeUInt32().createColumnConst(1, shards), std::make_shared<DataTypeUInt32>(), "_shard_count"}});
-    auto external_tables = context->getExternalTables();
-
-    if (!remote_shards.empty())
-    {
-        auto new_context = Context::createCopy(context);
-
-        for (const auto & shard : remote_shards)
-        {
-            auto read_from_remote = std::make_unique<ReadFromParallelRemoteReplicasStep>(
-                shard.coordinator,
-                shard,
-                shard.header,
-                processed_stage,
-                main_table,
-                table_func_ptr,
-                new_context,
-                throttler,
-                scalars,
-                external_tables,
-                &Poco::Logger::get("ReadFromParallelRemoteReplicasStep"),
-                query_info.storage_limits);
-
-            auto remote_plan = std::make_unique<QueryPlan>();
-            remote_plan->addStep(std::move(read_from_remote));
-            remote_plan->addInterpreterContext(new_context);
-            plans.emplace_back(std::move(remote_plan));
-        }
-    }
-
-    if (plans.empty())
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "No plans were generated for reading from Distributed. This is a bug");
-
-    if (plans.size() == 1)
+    if (not_optimized_cluster->getShardsInfo().size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cluster for parallel replicas should consist only from one shard");
+
+    auto shard_info = not_optimized_cluster->getShardsInfo().front();
+
+    if (auto it = stream_factory.objects_by_shard.find(shard_info.shard_num); it != stream_factory.objects_by_shard.end())
+        replaceMissedSubcolumnsByConstants(stream_factory.storage_snapshot->object_columns, it->second, query_ast);
+
+    auto all_replicas_count = shard_info.getAllNodeCount();
+    auto coordinator = std::make_shared<ParallelReplicasReadingCoordinator>(all_replicas_count);
+    auto remote_plan = std::make_unique<QueryPlan>();
+    auto plans = std::vector<QueryPlanPtr>();
+
+    /// This is a little bit weird, but we construct an "empty" coordinator without
+    /// any specified reading/coordination method (like Default, InOrder, InReverseOrder)
+    /// Because we will understand it later during QueryPlan optimization
+    /// So we place a reference to the coordinator to some common plane like QueryInfo
+    /// to then tell it about the reading method we chose.
+    query_info.coordinator = coordinator;
```

Comment:
Possible, start with Default and change to InOrder/InReverseOrder when optimization is applied?
But looks ok geenrally

Reply:
Coordinator class is just an interface. It stores an Impl object inside and creates it lazily. It doesn't make sense to create a Default Impl here.